### PR TITLE
BLUEBUTTON-1112: Bye bye JBoss

### PR DESCRIPTION
**Pull request by karlmdavis**
Remove JBoss/Wildfly and replace them with `bfd-server-launcher`: a small whitelabeled Jetty server.

The RFC for this architectural change is also part of this branch (along with the proposed implementation): [0000-bye-bye-jboss](https://github.com/CMSgov/beneficiary-fhir-data/blob/a5a5ee465d4e0eab7a13b03707cce7b9e42e4bdd/rfcs/0000-jboss-replacement.md).

In general, I tried to keep things reasonably backwards compatible. It's worth noting that the existing ITs all run unchanged.

The following differences were intentional:

1. The access logs have slightly different paths (same format, though).
2. There is no text message log anymore; it's NDJSON only.
3. There is no equivalent to the JBoss `server.log` file: all of the container logging is rolled into the normal NDJSON "message" log.
4. The application binary itself writes its NDJSON logs out to `STDOUT` rather than a file. The service launch script redirects this to a file.
    * This is a more "12" factor approach (though the access logs are written to separate files).
5. There's no log rolling anymore.
6. There's no DB query log anymore.
    * This was mostly a failed experiment and not worth carrying over.

The following things were intentionally kept the same for now, but should be changed/fixed soon:

1. The server still returns SSL handshake errors rather than HTTP `401`s/`403`s for authentication issues.
2. The NDJSON access logs are still using the same format, which appears to be incompatible with AWS insights.

### CODE REVIEW

**Comment by karlmdavis** [comment]
@cthayes, @cthulhuplus, @dho937, @jcgeer, @jeskew-gov, @jzulim, @njdister, @RickHawesUSDS 

Could I get a review on the RFC here?

The implementation is _mostly_ done, but not quite there yet, so please hold off there.

**Comment by karlmdavis** [comment]
Based on the feedback I've received, I believe this proposal has achieved wide consensus. Am I right? Let's find out!

I'm making an official "motion for final comment," with a disposition to **approve/merge** the RFC embedded in this PR: [0000-bye-bye-jboss](https://github.com/CMSgov/beneficiary-fhir-data/blob/a5a5ee465d4e0eab7a13b03707cce7b9e42e4bdd/rfcs/0000-jboss-replacement.md).

Please **add a reaction to this comment** to vote:

* If you agree, please give it a 👍.
* If you disagree, give it a 👎.
* If you'd like to officially abstain, give it a 🦆.

The following folks are directly impacted by this proposal, and should vote on it: @karlmdavis, @dho937, @jcgeer, @cthayes, @ibh1127, @jzulim, @RickHawesUSDS, @njdister, @cthulhuplus, @jeskew-gov.

File: apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+   * @param envVarKey the name of the environment variable to read
+   * @return the specified environment variable's value, as a {@link Path}
+   */
+  private static Path readEnvVarAsPath(String envVarKey) {
+    String pathText = System.getenv(envVarKey);
+    if (pathText == null || pathText.isEmpty())
+      throw new AppConfigurationException(
+          String.format("Missing value for configuration environment variable '%s'.", envVarKey));
+    Path path = Paths.get(pathText);
+    return path;
```
Comment (jeskew-gov) [inline apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java]: As above, I think this is easier to follow as an optional chain:

```
return Optional.ofNullable(envVarKey)
        .filter(v -> !v.isEmpty())
        .map(Paths::get)
        .orElseThrow(...);
```

File: ops/ansible/playbooks-healthapt/templates/awslogs.conf.data-server.j2 [hunk 2]
```diff
@@ -135,60 +135,36 @@ log_stream_name = {instance_id}
 initial_position = start_of_file
 log_group_name = /var/log/secure
 
-[Jboss_Access_Log]
+[bfd_server_access_txt]
```
Comment (jeskew-gov) [inline ops/ansible/playbooks-healthapt/templates/awslogs.conf.data-server.j2]: You may need to rebase and make these modifications to the CW unified agent JSON configuration instead. Sorry!

File: ops/ansible/roles/bfd-server/files/bluebutton-appserver-config.sh [hunk 3]
```diff
@@ -1,336 +0,0 @@
-#!/bin/bash
-
```
Comment (jeskew-gov) [inline ops/ansible/roles/bfd-server/files/bluebutton-appserver-config.sh]: Very happy to see this file go. Failures within this script were not fun to debug.

File: ops/ansible/roles/bfd-server/tasks/main.yml [hunk 4]
```diff
@@ -1,16 +1,18 @@
 ---
 
-- include: base_install.yml
+# Ensure that the service isn't started/restarted unless the 'post-ami' tag is active.
+# Note: this must always be referenced as `launch_enabled | default(false)`.
+- name: Activate Launch Tasks If post-ami Tag Is Active
+  set_fact:
+    launch_enabled: true
+  tags:
+    post-ami
```
Comment (jeskew-gov) [inline ops/ansible/roles/bfd-server/tasks/main.yml]: `    - post-ami` (need the dash to make `tags` an array)

File: ops/ansible/playbooks-ccs/build_bfd-server.yml [hunk 5]
```diff
@@ -45,7 +42,7 @@
         data_server_db_url: "{{ vault_data_db_secondary_url }}"
         data_server_db_username: "{{ vault_data_server_db_username }}"
         data_server_db_password: "{{ vault_data_server_db_password }}"
-        data_server_db_connections_max: 40
+        data_server_db_max_connections_max: 40
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply ops/ansible/playbooks-ccs/build_bfd-server.yml]: Pretty sure I goofed something here -- that doesn't look correct to me. I'll investigate -- thanks for catching!

File: apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+/** Models the configuration options for the launcher. */
+public final class AppConfiguration implements Serializable {
+  private static final long serialVersionUID = -8134110339292440841L;
+
+  /**
+   * The name of the environment variable that should be used to provide the {@link #getPort()}
+   * value.
+   */
+  public static final String ENV_VAR_KEY_PORT = "PORT";
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java]: Nice suggestion.

File: apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      throw new AppConfigurationException(
+          String.format("Missing value for configuration environment variable '%s'.", envVarKey));
+    int intValue;
+    try {
+      intValue = Integer.parseInt(intText);
+    } catch (NumberFormatException e) {
+      throw new AppConfigurationException(
+          String.format("Invalid value for configuration environment variable '%s'.", envVarKey));
+    }
+    return intValue;
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/AppConfiguration.java]: I'll have to stare at this a bit to make sure the semantics all still match, but I like the idea. I wasn't aware that `Optional` supported `filter`, `map`, etc. -- interesting!

File: apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/DataServerLauncherApp.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    // Create the HTTPS config.
+    HttpConfiguration httpsConfig = new HttpConfiguration(httpConfig);
+    httpsConfig.addCustomizer(new SecureRequestCustomizer());
+
+    // Create the SslContextFactory to be used, along with the cert.
+    SslContextFactory.Server sslContextFactory = new SslContextFactory.Server();
+    sslContextFactory.setNeedClientAuth(true);
+    sslContextFactory.setKeyStoreResource(Resource.newResource(appConfig.getKeystore()));
+    sslContextFactory.setCertAlias("server");
+    sslContextFactory.setKeyManagerPassword("changeit");
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/DataServerLauncherApp.java]: If I could disable encryption for these I would -- just don't find it adds much value in 90% of the cases, where you're going to have the password somewhere on the same disk in plaintext.

File: ops/ansible/roles/bfd-server/files/bluebutton-appserver-config.sh (hunk 3 above)
Comment (karlmdavis) [reply ops/ansible/roles/bfd-server/files/bluebutton-appserver-config.sh]: Same!

Hated the thing from the moment I wrote it.

File: apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/DataServerLauncherApp.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+   * @param metrics the {@link MetricRegistry} to log out before the application exits
+   */
+  private static void registerShutdownHook(MetricRegistry metrics) {
+    Runtime.getRuntime()
+        .addShutdownHook(
+            new Thread(
+                new Runnable() {
+                  @Override
+                  public void run() {
+                    LOGGER.info("Application is shutting down. Housekeeping...");
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-launcher/src/main/java/gov/cms/bfd/server/launcher/DataServerLauncherApp.java]: Why here, specifically?

**Pull request merged**